designs/autoscaler.md

```diff
@@ -0,0 +1,526 @@
+    name: md-0
+  - count: 1
+    autoscalingConfiguration:
+    - minCount: 1
+    - maxCount: 3
+    machineGroupRef:
+      kind: VSphereMachineConfig
+      name: worker-machine-b
+    name: md-1
+...
+```
+
+Make sure cluster-autoscaler is available to install as an EKS-A package:
+
+```sh
+$ eksctl anywhere list packages --source registry --kube-version 1.21
+
+// Example command output
+Package                 Version(s)
+-------                 ----------
+cluster-autoscaler      1.0.0-164563266364sh4209000f3dgfea84eed36529sa
+harbor                  2.5.0-4324383d8c5383bded5f7378efb98b4d50af827b
+...
+```
+
+Generate a curated-package config for autoscaler and run create command with it plus the cluster spec:
+
+```sh
+$ eksctl anywhere generate package cluster-autoscaler --source registry -f cluster.yaml > cas-package.yaml
+```
+
+a) enables autoscaling in management cluster, where the autoscaler controller is deployed in management cluster and watches the management cluster itself.
+
+```yaml
+// cas-package.yaml
+apiVersion: packages.eks.amazonaws.com/v1alpha1
+kind: Package
+metadata:
+  name: cluster-autoscaler-management-cluster
+  namespace: eksa-packages
+spec:
+  packageName: cluster-autoscaler
+  config:
+    cloudProvider: "clusterapi"
+    autoDiscovery:
+      clusterName: "management-cluster"
+```
+
+```sh
+$ eksctl anywhere create cluster -f mgmt-cluster.yaml --install-packages cas-package.yaml`
+```
+
+This command goes through the following steps:
+
+1. create a new EKS-A cluster with proper autoscaling annotations on worker resources
+2. install curated package controller on the management cluster
+3. the package controller on the management cluster then installs the Cluster Autoscaler package on management cluster
+
+b) enables autoscaling in workload cluster, where the autoscaler controller is deployed in workload cluster and watches the workload cluster itself.
+
+```yaml
+// cas-package.yaml
+apiVersion: packages.eks.amazonaws.com/v1alpha1
+kind: Package
+metadata:
+  name: cluster-autoscaler-workload-cluster
+  namespace: eksa-packages
+spec:
+  packageName: cluster-autoscaler
+  config:
+    cloudProvider: "clusterapi"
+    autoDiscovery:
+      clusterName: "workload-cluster"
+    clusterAPIMode: "incluster-kubeconfig"
+    clusterAPICloudConfigPath: "/etc/kubernetes/value"
+    extraVolumeSecrets:
+      cluster-autoscaler-cloud-config:
+        mountPath: "/etc/kubernetes"
+        name: "management-cluster-kubeconfig"
+```
+
+```sh
+$ eksctl anywhere create cluster -f workload-cluster.yaml --install-packages cas-package.yaml --kubeconfig mgmt-cluster.kubeconfig
+```
+
+This command goes through the following steps:
+
+1. create a new EKS-A cluster with proper autoscaling annotations on worker resources
+2. install curated package controller on the management cluster
+3. store the management cluster’s kubeconfig as a secret in the workload cluster (through curated package webhook, TBD)
```

Comment:
Do we need to bake this into our product? I'm not sure we have to, it seems like extra work that can be avoided. Imo we shouldn't recommend this option at all, let alone build it.

If in the future there are enough users who want this, we could add it. But for the first iteration I think it's best to only provide as an option our preferred topology.

Also, if we go with this, we need to take into consideration the security implications of adding that capability to the curated packages controller. In order to create that kubeconfig, the controller needs access to read it, which gives it admin access. There is a big difference between just having access to create deployments (or whatever else) and having admin access.

Reply:
Imo we should, because there is no way for us to prevent user from installing autoscaler in workload cluster themselves, even through curated package. To me it is a choice made based on different use cases: resource allocation, infrastructure and deployment strategy, etc. We tell the user pros and cons for each approach, they determine which option they go with. Plus I believe we will support creating cluster with installing pkg on the cluster itself first. Treating autoscaler as a package to me means user has the flexibility to choose how and where to install this package. We can have opinionated suggestion, but also provide instructions for each option that is durable.

Reply:
I see what you are saying, however, there is a difference between providing instructions to do something and baking everything into the code.

The more functionality we add to the codebase, the more we need to maintain. And that's a fee we need to pay for life. Hence my preference to start simple and add more features as users request them.